8299329: Assertion failure with fastdebug build when trying to use CDS without classpath

ashu-mehra opened the pull request:
It appears longest_common_app_classpath_prefix_len() is not returning correct value when there is no file separator in the path being searched backwards. Instead of returning 0 it return 1.

In case of empty classpath, it can result in assertion failure in check_paths():

`assert(strlen(rp_array->at(i)) > (size_t)runtime_prefix_len, "sanity");`

It can also result in incorrectly validating the app classpaths if they only differ by first character. Eg:

Dump time:
  -cp hello.jar
Run time:
  -cp mello.jar

This would not result in classpath mismatch!

This fix updates longest_common_app_classpath_prefix_len() to return 0 if no file separator character is found.

Signed-off-by: Ashutosh Mehra <asmehra@redhat.com>

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8299329](https://bugs.openjdk.org/browse/JDK-8299329): Assertion failure with fastdebug build when trying to use CDS without classpath


### Reviewers
 * [Ioi Lam](https://openjdk.org/census#iklam) (@iklam - **Reviewer**) ⚠️ Review applies to [e4fec7e7](https://git.openjdk.org/jdk/pull/11781/files/e4fec7e703b695e6e7338f5cda6db83f68ea65da)
 * [Calvin Cheung](https://openjdk.org/census#ccheung) (@calvinccheung - **Reviewer**) ⚠️ Review applies to [e4fec7e7](https://git.openjdk.org/jdk/pull/11781/files/e4fec7e703b695e6e7338f5cda6db83f68ea65da)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11781/head:pull/11781` \
`$ git checkout pull/11781`

Update a local copy of the PR: \
`$ git checkout pull/11781` \
`$ git pull https://git.openjdk.org/jdk pull/11781/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11781`

View PR using the GUI difftool: \
`$ git pr show -t 11781`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11781.diff">https://git.openjdk.org/jdk/pull/11781.diff</a>

</details>

jerboaa commented:
Could we add a test for this a la `test/hotspot/jtreg/runtime/cds/appcds/WrongClasspath.java`?

ashu-mehra commented:
@jerboaa I see WrongClasspath.java already has a test for "no classpath" condition, but it is not failing with debug build.
The command line used for running "no classpaht" test reveals it is actually running with a classpath:

```
/work/jdk/build/base-fd/images/jdk/bin/java -cp /work/jdk/JTwork/classes/runtime/cds/appcds/WrongClasspath.d:/work/jdk/test/hotspot/jtreg/runtime/cds/appcds:/work/jdk/JTwork/classes/test/lib:/work/jdk/test/lib:/work/jtreg/build/images/jtreg/lib/javatest.jar:/work/jtreg/build/images/jtreg/lib/jtreg.jar -Xshare:on -showversion -XX:SharedArchiveFile=/work/jdk/JTwork/scratch/appcds-15h18m10s791.jsa -Dtest.timeout.factor=1.0 -XX:+UnlockDiagnosticVMOptions -XX:VerifyArchivedFields=1 -Xlog:cds Hello
```

The classpath is appended by jdk.test.lib.process.ProcessTools#createJavaProcessBuilder():

```
        args.add("-cp");
        args.add(System.getProperty("java.class.path"));
```

That means the "no classpath" test is not really testing what it is expected to.

iklam commented:
> @jerboaa I see WrongClasspath.java already has a test for "no classpath" condition, but it is not failing with debug build. The command line used for running "no classpaht" test reveals it is actually running with a classpath:
> 
> ```
> /work/jdk/build/base-fd/images/jdk/bin/java -cp /work/jdk/JTwork/classes/runtime/cds/appcds/WrongClasspath.d:/work/jdk/test/hotspot/jtreg/runtime/cds/appcds:/work/jdk/JTwork/classes/test/lib:/work/jdk/test/lib:/work/jtreg/build/images/jtreg/lib/javatest.jar:/work/jtreg/build/images/jtreg/lib/jtreg.jar -Xshare:on -showversion -XX:SharedArchiveFile=/work/jdk/JTwork/scratch/appcds-15h18m10s791.jsa -Dtest.timeout.factor=1.0 -XX:+UnlockDiagnosticVMOptions -XX:VerifyArchivedFields=1 -Xlog:cds Hello
> ```
> 
> The classpath is appended by jdk.test.lib.process.ProcessTools#createJavaProcessBuilder():
> 
> ```
>         args.add("-cp");
>         args.add(System.getProperty("java.class.path"));
> ```
> 
> That means the "no classpath" test is not really testing what it is expected to.

Maybe we can do this?

```
    // Then try to execute the archive without -classpath -- it should fail
    TestCommon.run(
-       /* "-cp", appJar, */ // <- uncomment this and the execution should succeed
+       // Force classpath to be the empty string. Note: -cp "" doesn't work on Windows.
+       "-Djava.class.path=",
        "-Xlog:cds",
        "Hello")
        .assertAbnormalExit(unableToUseMsg, mismatchMsg, hintMsg);
```

That should override the `-cp` parameter given by `createJavaProcessBuilder`, which appears earlier in the command-line.

Also, I think you should add the `-cp hello.jar` vs `-cp mello.jar` test case into WrongClasspath.java.

Thanks!

ashu-mehra commented:
Adding `-Djava.class.path=` wouldn't work as it is different than the case when no classpath is specified. When there is no classpath specified, launcher adds "." as the classpath.
To fix the "no-classpath" test I added another system property `test.noclasspath` to avoid setting 'cp' in `ProcessTools#createJavaProcessBuilder`. I realized the CLASSPATH env variable is also set when the test Java process is created, so that needs to be removed as well.

> Also, I think you should add the -cp hello.jar vs -cp mello.jar test case into WrongClasspath.java.

I have added this test as well.

iklam commented on test/hotspot/jtreg/runtime/cds/appcds/WrongClasspath.java:
```diff
@@ -40,14 +40,29 @@ public class WrongClasspath {
+    // Dump CDS archive with hello.jar
+    // Run with a jar file that differs from the original jar file by the first character only: -cp mello.jar
+    // Shared class paths mismatch should be detected.
+    String mellojar = appJarInWorkDir.replace("hello.jar", "mello.jar");
```
I am not sure if we need to add more APIs to JarBuilder. Since you just need two files "hello.jar" and "mello.jar", I think that can be done like:

```
import java.nio.file.Paths;
...
String helloJar = "hello.jar";
String melloJar = "mello.jar";
Files.copy(Paths.get(appJar), Paths.get(helloJar), StandardCopyOption.COPY_ATTRIBUTES);
Files.copy(Paths.get(appJar), Paths.get(melloJar), StandardCopyOption.COPY_ATTRIBUTES);
```

ashu-mehra replied:
yup, that would do as well. Pushed a commit as per the suggestion.

iklam reviewed (approved):
LGTM

calvinccheung commented on src/hotspot/share/cds/filemap.cpp:
```diff
@@ -880,11 +880,13 @@ unsigned int FileMapInfo::longest_common_app_classpath_prefix_len(int num_paths,
-      return pos + 1;
+      while (pos > 0) {
+        if (rp_array->at(0)[--pos] == *os::file_separator()) {
+          return pos+1;
```
Please add a space before and after the '+'.

calvinccheung reviewed (approved):
Thanks for fixing this. Just one nit below.

ashu-mehra commented:
There is a test failure on macos-x64 not related to the changes in this PR.
The test `test/hotspot/jtreg/gc/g1/TestVerifyGCType.java` fails at:

```
java.lang.RuntimeException: Missing expected verification pattern Verifying After GC for: Pause Young (Prepare Mixed): expected true, was false
	at jdk.test.lib.Asserts.fail(Asserts.java:594)
	at jdk.test.lib.Asserts.assertTrue(Asserts.java:486)
	at gc.g1.TestVerifyGCType.verifyType(TestVerifyGCType.java:201)
	at gc.g1.TestVerifyGCType.verifyCollection(TestVerifyGCType.java:196)
	at gc.g1.TestVerifyGCType.testYoungEvacFail(TestVerifyGCType.java:134)
	at gc.g1.TestVerifyGCType.main(TestVerifyGCType.java:57)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
	at java.base/java.lang.reflect.Method.invoke(Method.java:578)
	at com.sun.javatest.regtest.agent.MainActionHelper$AgentVMRunnable.run(MainActionHelper.java:312)
	at java.base/java.lang.Thread.run(Thread.java:1623)
```

Looks like this is same as [JDK-8298215](https://bugs.openjdk.org/browse/JDK-8298215) which is already fixed. I guess the branch is not up-to-date and doesn't have the required fix, which is why this test failed. I think the failure can be ignored.

ashu-mehra commented:
/sponsor

ashu-mehra commented:
/integrate

calvinccheung commented:
> There is a test failure on macos-x64 not related to the changes in this PR. The test `test/hotspot/jtreg/gc/g1/TestVerifyGCType.java` fails at:
> 
> ```
> java.lang.RuntimeException: Missing expected verification pattern Verifying After GC for: Pause Young (Prepare Mixed): expected true, was false
> 	at jdk.test.lib.Asserts.fail(Asserts.java:594)
> 	at jdk.test.lib.Asserts.assertTrue(Asserts.java:486)
> 	at gc.g1.TestVerifyGCType.verifyType(TestVerifyGCType.java:201)
> 	at gc.g1.TestVerifyGCType.verifyCollection(TestVerifyGCType.java:196)
> 	at gc.g1.TestVerifyGCType.testYoungEvacFail(TestVerifyGCType.java:134)
> 	at gc.g1.TestVerifyGCType.main(TestVerifyGCType.java:57)
> 	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
> 	at java.base/java.lang.reflect.Method.invoke(Method.java:578)
> 	at com.sun.javatest.regtest.agent.MainActionHelper$AgentVMRunnable.run(MainActionHelper.java:312)
> 	at java.base/java.lang.Thread.run(Thread.java:1623)
> ```
> 
> Looks like this is same as [JDK-8298215](https://bugs.openjdk.org/browse/JDK-8298215) which is already fixed. I guess the branch is not up-to-date and doesn't have the required fix, which is why this test failed. I think the failure can be ignored.

Yes, if your repo doesn't contain the fix for JDK-8298215, the failure could be ignored.

calvinccheung commented:
/sponsor

dholmes-ora commented:
This change has introduced a test failure in our tier 4 testing:

https://bugs.openjdk.org/browse/JDK-8299699
